File: examples/Ruby/anpr/anpr.rb
=begin
The MIT License (MIT)
Copyright © 2022-2025 TS-Solution Corp.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to all conditions.

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
=end

require 'rbconfig'
require 'fileutils'
require 'ffi'
require 'vips'

# Set base directory
EXAMPLES_BASE_DIR = File.expand_path(File.join(__dir__, "../.."))

# Determine native library path based on OS and architecture
def get_engine_file_name
  arch = RbConfig::CONFIG['host_cpu']
  os = RbConfig::CONFIG['host_os']

  if os =~ /mswin|mingw|cygwin/
    if arch =~ /x64|amd64|x86_64/
      return File.join(EXAMPLES_BASE_DIR, 'bin/windows-x86_64/tsanpr.dll')
    elsif arch =~ /x86|i386/
      return File.join(EXAMPLES_BASE_DIR, 'bin/windows-x86/tsanpr.dll')
    end
  elsif os =~ /linux/
    if arch =~ /x86_64|amd64/
      return File.join(EXAMPLES_BASE_DIR, 'bin/linux-x86_64/libtsanpr.so')
    elsif arch =~ /aarch64/
      return File.join(EXAMPLES_BASE_DIR, 'bin/linux-aarch64/libtsanpr.so')
    end
  end
  ""
end

# Set native library path as environment variable
engine_file_name = get_engine_file_name
if engine_file_name.empty? || !File.exist?(engine_file_name)
  puts "Unsupported operating system or engine not found"
  exit(1)
end
ENV['TSANPR_LIB_PATH'] = engine_file_name

require_relative './lib/tsanpr'

# call anpr_read_file with image file name
def read_image_file(tsanpr, imgfile, output_format, options)
  print "#{imgfile} (outputFormat=\"#{output_format}\", options=\"#{options}\") => "
  result = tsanpr.anpr_read_file(imgfile, output_format, options)
  puts result
end

# Read an encoded image file (e.g., JPG/PNG) as a binary buffer and call anpr_read_pixels with "encoded" pixel format
def read_encoded_image(tsanpr, imgfile, output_format, options)
  print "#{imgfile} (outputFormat=\"#{output_format}\", options=\"#{options}\") => "
  begin
    encoded_img = File.binread(imgfile)
    # Create an FFI::MemoryPointer from the binary data
    FFI::MemoryPointer.new(:uint8, encoded_img.bytesize) do |buf|
      buf.put_bytes(0, encoded_img)
      # Call the wrapper function defined in tsanpr.rb
      result = tsanpr.anpr_read_pixels(
        buf, encoded_img.bytesize, 0, 0, "encoded", output_format, options
      )
      puts result
    end
  rescue => e
    puts "\nERROR: Exception - #{e.message}"
  end
end

# Read image file into pixel buffer and call anpr_read_pixels
def read_pixel_buffer(tsanpr, imgfile, output_format, options)
  print "#{imgfile} (outputFormat=\"#{output_format}\", options=\"#{options}\") => "
  begin
    image = Vips::Image.new_from_file(imgfile, access: :sequential)
    width = image.width
    height = image.height
    bands = image.bands # 3 for RGB, 4 for RGBA, 1 for GRAY, etc.

    # Determine pixel format string based on bands
    pixel_format =
      case bands
      when 1
        "GRAY"
      when 3
        "RGB"
      when 4
        "RGBA"
      else
        raise "Unsupported number of bands: #{bands}"
      end

    # Get raw pixel buffer (binary string)
    pixels = image.write_to_memory
    stride = width * bands

    # If BGR is needed, swap channels here
    FFI::MemoryPointer.new(:uint8, pixels.bytesize) do |buf|
      buf.put_bytes(0, pixels)
      result = tsanpr.anpr_read_pixels(buf, width, height, stride, pixel_format, output_format, options)
      puts result
    end
  rescue => e
    puts "\nERROR: Exception - #{e.message}"
  end
end

# Run license plate recognition for a country
def read_license_plates(tsanpr, country_code)
  # NOTICE:
  # anpr_initialize should be called only once after library load.
  # Therefore, it is not possible to change the country code after anpr_initialize has been called.
  # While using the free trial license, you can try all languages.
  # When you purchase a commercial license, you can only use the selected language.
  error = tsanpr.anpr_initialize("text;country=#{country_code}")
  unless error.empty?
    puts "anpr_initialize() failed: #{error}"
    return
  end
  
  image_dir = File.join(EXAMPLES_BASE_DIR, "img", country_code)

  # TODO:  Try each function as needed
  read_func = method(:read_image_file)
  #read_func = method(:read_encoded_image)
  # read_func = method(:read_pixel_buffer)

  # TODO: Try each output format as needed
  output_format = "text"
  # output_format = "json"
  # output_format = "yaml"
  # output_format = "xml"
  # output_format = "csv"

  read_func.call(tsanpr, File.join(image_dir, "licensePlate.jpg"), output_format, "") # Single license plate recognition (default)
  read_func.call(tsanpr, File.join(image_dir, "multiple.jpg"), output_format, "vm")   # Recognize multiple license plates attached to vehicles
  read_func.call(tsanpr, File.join(image_dir, "multiple.jpg"), output_format, "vmb")  # Recognize multiple license plates attached to vehicles (including motorcycles)
  read_func.call(tsanpr, File.join(image_dir, "surround.jpg"), output_format, "vms")  # Recognize multiple license plates attached to vehicles with surround detection
  read_func.call(tsanpr, File.join(image_dir, "surround.jpg"), output_format, "dms")  # Recognize multiple surrounding objects (vehicles)
  read_func.call(tsanpr, File.join(image_dir, "surround.jpg"), output_format, "dmsr") # Recognize multiple surrounding objects (vehicles) and license plates
end

def main
  tsanpr = TSANPR.new

  # TODO: Try each country code as needed
  read_license_plates(tsanpr, "KR")
  # read_license_plates(tsanpr, "JP")
  # read_license_plates(tsanpr, "VN")
end

main if __FILE__ == $0

File: examples/Perl/anpr/anpr.pl
#!/usr/bin/env perl
#
# The MIT License (MIT)
# Copyright © 2022-2025 TS-Solution Corp.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to all conditions.
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

use strict;
use warnings;
use Config;
use Imager;
use lib './lib';
use TSANPR qw(load_tsanpr unload_tsanpr anpr_initialize anpr_read_file anpr_read_pixels);
use FFI::Platypus::Buffer qw(scalar_to_pointer);

my $examplesBaseDir = "../..";

sub getEngineFileName {
    my $base = $examplesBaseDir;
    if ($^O eq 'MSWin32') {
        # 64-bit or 32-bit Windows
        if ($Config::Config{archname} =~ /x64/) {
            return $base . "\\bin\\windows-x86_64\\tsanpr.dll";
        } else {
            return $base . "\\bin\\windows-x86\\tsanpr.dll";
        }
    } elsif ($^O eq 'linux') {
        my $arch = `uname -m`;
        chomp $arch;
        return $base . "/bin/linux-x86_64/libtsanpr.so" if $arch eq 'x86_64';
        return $base . "/bin/linux-aarch64/libtsanpr.so" if $arch eq 'aarch64';
    }
    die "Unsupported OS or architecture";
}

sub readImageFile {
    my ($imgfile, $outputFormat, $options) = @_;
    print "$imgfile (outputFormat=\"$outputFormat\", options=\"$options\") => ";
    my $result = anpr_read_file($imgfile, $outputFormat, $options);
    print "$result\n";
}

sub readEncodedImage {
    my ($imgfile, $outputFormat, $options) = @_;
    print "$imgfile (outputFormat=\"$outputFormat\", options=\"$options\") => ";
    open my $fh, '<:raw', $imgfile or do {
        print "File open failed\n";
        return;
    };
    my $data;
    read $fh, $data, -s $fh;
    close $fh;

    # Get pointer to the data
    my $ptr = scalar_to_pointer($data);

    my $result = anpr_read_pixels($ptr, length($data), 0, 0, "encoded", $outputFormat, $options);
    print "$result\n";
}

sub readPixelBuffer {
    my ($imgfile, $outputFormat, $options) = @_;
    print "$imgfile (outputFormat=\"$outputFormat\", options=\"$options\") => ";

    # Read the image file
    my $img = Imager->new;
    $img->read(file => $imgfile)
        or die "Cannot read $imgfile: ", $img->errstr;

    # Extract image information
    my $width  = $img->getwidth;
    my $height = $img->getheight;
    my $channels = $img->getchannels;

    # Determine pixel format
    my $pixelFormat;
    if ($channels == 1) {
        $pixelFormat = "GRAY";
    } elsif ($channels == 3) {
        $pixelFormat = "BGR"; # Imager uses RGB by default; convert if necessary
    } elsif ($channels == 4) {
        $pixelFormat = "BGRA";
    } else {
        die "Unsupported channel count: $channels";
    }

    # Extract raw pixel data
    my $raw;
    $img->write(data => \$raw, type => 'raw')
        or die "Cannot extract raw pixel data: ", $img->errstr;

    # Calculate stride (bytes per row)
    my $stride = $width * $channels;

    # Convert raw data to pointer
    my $ptr = scalar_to_pointer($raw);

    # Call anpr_read_pixels with pointer
    my $result = anpr_read_pixels($ptr, $width, $height, $stride, $pixelFormat, $outputFormat, $options);
    print "$result\n";
}

sub readLicensePlates {
    # NOTICE:
	# anpr_initialize should be called only once after library load.
	# Therefore, it is not possible to change the country code after anpr_initialize has been called.
	# While using the free trial license, you can try all languages.
	# When you purchase a commercial license, you can only use the selected language.
    my ($countryCode) = @_;
    my $initParams = "text;country=$countryCode";
    my $error = anpr_initialize($initParams);
    if ($error && length($error)) {
        print "anpr_initialize() failed (error=$error)\n";
        return -1;
    }
    my $imageDir = "$examplesBaseDir/img/$countryCode/";

    # TODO: Try each function as needed
    my $anprFunc = \&readImageFile;
    # my $anprFunc = \&readEncodedImage;
    # my $anprFunc = \&readPixelBuffer;

    # TODO: Try each output format as needed
    my $outputFormat = "text";
    # my $outputFormat = "json";
    # my $outputFormat = "yaml";
    # my $outputFormat = "xml";
    # my $outputFormat = "csv";

    $anprFunc->("${imageDir}licensePlate.jpg", $outputFormat, "");  # Single license plate recognition (default)
    $anprFunc->("${imageDir}multiple.jpg", $outputFormat, "vm");    # Recognize multiple license plates attached to vehicles
    $anprFunc->("${imageDir}multiple.jpg", $outputFormat, "vmb");   # Recognize multiple license plates attached to vehicles (including motorcycles)
    $anprFunc->("${imageDir}surround.jpg", $outputFormat, "vms");   # Recognize multiple license plates attached to vehicles with surround detection
    $anprFunc->("${imageDir}surround.jpg", $outputFormat, "dms");   # Recognize multiple surrounding objects (vehicles)
    $anprFunc->("${imageDir}surround.jpg", $outputFormat, "dmsr");  # Recognize multiple surrounding objects (vehicles) and license plates

    # Recognize multiple surrounding objects and license plates within RoI
    $anprFunc->("${imageDir}surround.jpg", $outputFormat, "dmsri549,700,549,2427,1289,2427,1289,700")
    
    return 0;
}

# Main
my $engineFileName = getEngineFileName();
load_tsanpr($engineFileName);

# TODO: Try each country code as needed
readLicensePlates("KR");
# readLicensePlates("JP");
# readLicensePlates("VN");

unload_tsanpr();
